background/index.ts

```diff
@@ -94,6 +97,172 @@ async function main() {
     { pendingTransaction: {} }
   )
 
+  const indexBalanceOnBlock = async () => {
+    const blockSubscriberContext: {
+      provider?: JsonRpcProvider
+      blockSubscriber?: Listener
+    } = {}
+
+    // Syncs account balances using multicall
+    const updateAccountBalances = async (
+      accounts: Account[],
+      provider: JsonRpcProvider
+    ) => {
+      // Use multicall provider to optimize RPC calls
+      const multicallProvider = MulticallWrapper.wrap(provider)
+
+      // Store balance update promises
+      const tokenQueries = []
+
+      // Update balances for each account
+      accounts.forEach((account) => {
+        const {
+          id,
+          tokens,
+          chainId,
+          balance: accountBalance,
+          address: accountAddress
+        } = account
+
+        tokenQueries.push(
+          (async () => {
+            const nativeBalance =
+              await multicallProvider.getBalance(accountAddress)
+
+            // Update balance if changed
+            if (number.toBigInt(accountBalance) !== nativeBalance) {
+              storage.set((state) => {
+                state.account[id].balance = number.toHex(nativeBalance)
+              })
+
+              console.log(
+                `[background][indexBalanceOnBlock] Native token balance updated: ${nativeBalance.toString()} at chain ID: ${chainId}`
+              )
+            }
+          })()
+        )
+
+        tokens.forEach((t) => {
+          tokenQueries.push(
+            (async () => {
+              const tokenContract = await ERC20Contract.init(
```

Comment:
nit: I found `init` function should not be async

```suggestion
              const tokenContract = ERC20Contract.init(
```

Reply:
Modified in commit [https://github.com/pilagod/waallet-extension/pull/202/commits/f98b9d4b44ddf229d75dbc1264db808b75070aca](https://github.com/pilagod/waallet-extension/pull/202/commits/f98b9d4b44ddf229d75dbc1264db808b75070aca).

It has been observed that on the local network, there is an issue retrieving test ERC20 tokens during block mints.
Further investigation is needed to resolve this. 
```
Error: could not decode result data (value="0x", info={ "method": "tryAggregate", "signature": "tryAggregate(bool,(address,bytes)[])" }, code=BAD_DATA, version=6.13.1)
```